<!DOCTYPE html>
<html>
<body>

<div align="justify">
<p>For a quick test, select one example under "Step 1: data sets", or select "customData" and download the example annotated SVG file (aSVG, see below) file and data matrix in "Download". </p>

<p>This Shiny app is an integrated implementation of the R/Bioconductor package "spatialHeatmap". It is designed for interactively visualizing numeric values in biological assays (<em>e.g.</em> RNA-seq, microarray, qPCR) on an anotomical aSVG image. First, the core feature "Spatial Heatmap" maps numeric values of target assayed items (genes, proteins, metabolites, <em>etc</em>) to matching spatial features (cells, tissues, organs, <em>etc</em>) in the aSVG image. In this process, the values are translated into different colors and these colors are used to paint the spatial features. The resulting images are called spatial heatmaps (SHMs). Second, nearest neighbors of each target assayed item are selected by most similar expression profiles independently. All target items and their nearest neighbors are hierachically clustered and visualized as interactive matrix heatmap. Third, network modules are identified internally and the module containing a target item is displayed as interactive network graph. This app is a general visualization tool, not limited to biological data. In the following, instructions are given with a gene expression example. </p>

<p><b>When switching from "customData" or "customComputedData" to default examples, users should refresh the webpage before selecting new data set. Operations on this app are expected to follow the order as each panel appears. Otherwise if errors arise, the webpage should be refreshed.</b></p>
</div>

</body>
</html>







